Update docker-compose.yml

sontn1988 opened the pull request:
## Update info:
- Update docker-compose.yml: Update typesense version, update typesense use ENV file, disable public port 8108
- Update Readme
- Add typesense.env.default
- Add script start typesense

Commit:
Update docker-compose.yml

kenkunz commented on README.md:
````diff
@@ -15,7 +15,8 @@ self-hosted.
 
 ```bash
-export TYPESENSE_API_KEY=super_secret_ts_key  # replace with a secure key
+vim typesense.env
````
Instead of referencing `vim` here, I think it would be clearer to indicate (above the code block) that the ENV vars should be included in `typesense.env` file. I think the two separate code blocks could then be merged into one showing both options.

Let me know if you'd like me to draft some text for this.

sontn1988 replied:
I separate it to twe code blocks.
Pls help me to review it

kenkunz replied:
I added some additional documentation to reflect some changes I made to the setup.

kenkunz commented on docker-compose.yml:
```diff
@@ -1,11 +1,13 @@
       - ${TYPESENSE_DATA:-./typesense-data}:/data
+    env_file:
+      - $PWD/typesense.env
+    command: sh start_typesense.sh
```
This updated `command` line doesn't seem to work. The service fails to start, and keeps re-trying and giving the following error:
```
search-typesense-1 exited with code 1
```

This is because the `start_typesense.sh` script is _outside_ the container. The `command` option needs to specify something that can be executed _inside_ the container.

The [latest Typesense docs](https://typesense.org/docs/guide/install-typesense.html#docker-compose) show the following:

```
    command: '--data-dir /data --api-key=xyz --enable-cors'
```

When I change it to the above value, it works for me.

sontn1988 replied:
This issue because I had a mistake in the variable name
I fixed it in new commit

kenkunz replied:
This still wasn't working quite right.

The Typesense docker image uses the `ENTRYPOINT` instruction as the base command, so whatever is in the `CMD` instruction (passed from the `docker-compose.yml` `command` directive) is _appended_ to the entrypoint.

See:
- [Typesense deployment Dockerfile code](https://github.com/typesense/typesense/blob/master/docker/deployment.Dockerfile)
- [Typesense image layers / instructions](https://hub.docker.com/layers/typesense/typesense/0.25.2/images/sha256-59199e3c6a4952025f077e01599a532138082f78a72813bf8d1745c87c9bd5e5)
- [Explanation of how `ENTRYPOINT` AND `CMD` work together](https://kinsta.com/blog/dockerfile-entrypoint/#entrypoint-with-cmd)

Therefore… I modified the `command:` directive to follow the example in the [Typesense Docker Compose deployment guide](https://typesense.org/docs/guide/install-typesense.html#docker-compose).

See additional comments in 97c54a9d3d907de8f4c83b68b0bba94dd3a3f34e commit, as well as in `docker-compose.yml` and `README.md`.

kenkunz commented on start_typesense.sh:
Per my comment on `docker-compose.yml`, I don't think we can use this external script since we need a command to execute within the docker container.

kenkunz replied:
I verified that this script was not being invoked at all.

Since the `start_typesense.sh` command only exists _outside_ the container, it can't be invoked as a `command` (unless you were to first copy it into th container). Instead, the `/opt/typesense-server` command was being invoked _inside_ the container based on the `ENTRYPOINT` instruction (in the Typesense image). the `command` value was being appended to this, and since `sh start_typesense.sh` was unrecognized by the `typesense-server` command it was being ignored.

You can verify that `start_typesense.sh` was not being invoked…
1. check out 85e11ae5fa2d78cfcfabc84559a0bd3aaa8b0951
   ```bash
   git checkout 85e11ae
   ```
2. edit `start_typesense.sh` to replace `./typesense-server` line with… _anything_
3. make sure your env variables are configured
4. start the service
   ```bash
   docker-compose up
   ```

If `start_typesense.sh` was actually being invoked, this should fail… but it still works.

kenkunz commented on typesense.env.default:
I think the intention for this file is that someone can copy it to `typesense.env` and then configure it appropriately for their environment… correct?

If so, I think we should add `typesense.env` to `.gitignore` so someone doesn't accidentally check it in to the repo with real credentials.

Just curious… is there a reason to use a custom env file name `typesense.env` rather than the more typical / standard `.env`?

sontn1988 replied:
I added `typesense.env`  to `.gitignore`
I want to use env file name `typesense.env` rather than  `.env` because I want to distinguish the env file exactly for typesense

kenkunz replied:
As I noted in the `README`, it doesn't work quite right to use the `env_file` directive inside `docker-compose.yml`. Instead… if you want to name the env file `typesense.env`, you need to invoke `docker-compose` with the `--env-file` option. This behaves differently from the `env_file` directive:
- the `--env-file` command-line option (or a default `.env` file) makes the env vars available within the `docker-compose` file
- the `env_file` directive within `docker-compose` makes the env vars from the file available to processes running within the container.

We need the `TYPESENSE_HOST_DATA_DIR` var to be available in `docker-compose` itself (so we can use it in the `volumes` directive).

kenkunz reviewed (changes requested):
@sontn1988 sorry I wasn't able to get to this sooner.

See commends on specific files.

Commit:
Fix typesense.env variable

kenkunz committed:
update compose file command option and env vars

- `command` option should only include CLI flags since the default
  Typesense image ENTRYPOINT is `/opt/typesense-server`
- update env var names for self-host options to prevent conflicts with
  existing Typesense Cloud env vars (e.g., in secrets.env)

kenkunz committed:
remove unused `start_typesense.sh`

- this file only exists outside the container so it is not available to
  be run via the docker compose `command` option
- the `command` was being ignored since the value is appended to the
  docker image `ENTRYPOINT` of `/opt/typesense-server` (which ignored
  the unrecognized options)

kenkunz committed:
update env vars in template file; add comments

kenkunz committed:
update documentation with latest config info

kenkunz committed:
bump Typesense version to 0.25.2

see: https://github.com/typesense/typesense/releases/tag/v0.25.2

sontn1988 committed:
Update typesense.env.default to env.default

kenkunz committed:
minor tweaks to README and env.default

using ${ENV_VAR:-default_value} default syntax for consistency

kenkunz committed:
tweak README example
